protocols/relay/src/v2/relay.rs

```diff
@@ -0,0 +1,643 @@
+        vec![]
+    }
+
+    fn inject_connected(&mut self, _peer_id: &PeerId) {}
+
+    fn inject_disconnected(&mut self, _peer: &PeerId) {}
+
+    fn inject_connection_closed(
+        &mut self,
+        peer: &PeerId,
+        connection: &ConnectionId,
+        _: &ConnectedPoint,
+    ) {
+        self.reservations
+            .get_mut(peer)
+            .map(|cs| cs.remove(&connection))
+            .unwrap_or(false);
+
+        for circuit in self
+            .circuits
+            .remove_by_connection(*peer, *connection)
+            .iter()
+            // Only emit [`CircuitClosed`] for accepted requests.
+            .filter(|c| matches!(c.status, CircuitStatus::Accepted))
+        {
+            self.queued_actions
+                .push_back(NetworkBehaviourAction::GenerateEvent(
+                    Event::CircuitClosed {
+                        src_peer_id: circuit.src_peer_id,
+                        dst_peer_id: circuit.dst_peer_id,
+                        error: Some(std::io::ErrorKind::ConnectionAborted.into()),
+                    },
+                ));
+        }
+    }
+
+    fn inject_event(
+        &mut self,
+        event_source: PeerId,
+        connection: ConnectionId,
+        event: handler::Event,
+    ) {
+        match event {
+            handler::Event::ReservationReqReceived {
+                inbound_reservation_req,
+                remote_addr,
+            } => {
+                let now = Instant::now();
+
+                let handler_event = if remote_addr.iter().any(|p| p == Protocol::P2pCircuit) {
+                    // Deny reservation requests over relayed circuits.
+                    handler::In::DenyReservationReq {
+                        inbound_reservation_req,
+                        status: message_proto::Status::PermissionDenied,
+                    }
+                } else if self
+                    .reservations
+                    .iter()
+                    .map(|(_, cs)| cs.len())
+                    .sum::<usize>()
+                    >= self.config.max_reservations
+                    || !self
+                        .config
+                        .reservation_rate_limiters
+                        .iter_mut()
+                        .all(|limiter| limiter.try_next(event_source, &remote_addr, now))
+                {
+                    // Deny reservation exceeding limits.
+                    handler::In::DenyReservationReq {
+                        inbound_reservation_req,
+                        status: message_proto::Status::ResourceLimitExceeded,
+                    }
+                } else {
+                    // Accept reservation.
+                    self.reservations
+                        .entry(event_source)
+                        .or_default()
+                        .insert(connection);
+                    handler::In::AcceptReservationReq {
+                        inbound_reservation_req,
+                        addrs: vec![],
```

Comment:
here we construct an `AcceptReservationReq` with no addrs

Reply:
`addrs` is filled here:

https://github.com/libp2p/rust-libp2p/blob/2ef70508339cee3c9438bafc3986592aec5e20ab/protocols/relay/src/v2/relay.rs#L541-L556

This is definitely not intuitive, as you already noted. I don't see a cleaner way, without caching `poll_parameters.external_addresses`. Suggestions more than welcome.

Reply:
Does the below roughly follow your suggestion @MarcoPolo?

https://github.com/mxinden/rust-libp2p/commit/85b5a0882b91849e0c97379bcbd471ee57b8be5c

I like the idea, though the builder does not only have to carry the `inbound_reservation_req` but also the remote peer ID. I am not sure this is simpler than the current approach.

What do you think? I might misunderstand something.

Reply:
> .. also the remote peer ID.
Ah right I did forget that this is wrapped in `NetworkBehaviourAction::NotifyHandler`. 

I'm a little surprised the diff is net positive, but I guess that makes sense since you are now pushing the queued action in each if branch.

I think I still prefer https://github.com/mxinden/rust-libp2p/commit/85b5a0882b91849e0c97379bcbd471ee57b8be5c because:
1. In the current approach you create a struct that isn't complete yet, and rely on the developer to complete the struct.
2. In this approach you define a new type that can't be returned until it's been processed. You push the fact that you still have a step-to-do to the type system.

This is kind of similar to the idea of null vs option. Rust has an `Option` so that the type system forces you to handle the missing case.

The above is to share my thought process, but I don't feel to strongly here. I'll defer to you

Reply:
What do you think of https://github.com/mxinden/rust-libp2p/commit/031c891d9531e67dd059734c5f411a735cc8c788 as an alternative @MarcoPolo?

https://github.com/mxinden/rust-libp2p/commit/85b5a0882b91849e0c97379bcbd471ee57b8be5c (previous proposal) introduces a second queue (`queued_action_builders`). With two queues, total order of actions is no longer guaranteed. While I don't think any code within `libp2p-relay` depends on such total order, I still find it error prone and hard to debug.

https://github.com/mxinden/rust-libp2p/commit/031c891d9531e67dd059734c5f411a735cc8c788 (new proposal) uses a single queue only, thus upholding total order across all actions.

In my eyes the above is a workaround. In the long run I would like to redesign or get rid of the concept of `PollParamters`. I think problems like the above arise due to the mix of event driven (`inject_*`) and synchronous (`PollParameters`) pattern. See https://github.com/libp2p/rust-libp2p/discussions/2118#discussioncomment-1022115 for additional discussion.

Reply:
Clever! I think this upholds my preferences for relying on the type system.

Thanks for that thread link. Good to have context around PollParameters.

Reply:
Included via 112395275d9a9838fe802586090e1e2777e15f91. Let me know in case there is still something missing.